### PR TITLE
Fix the next-previous implementation to account for cursor position

### LINKED ISSUE #833: `Linter: Next Error` jumps to seemingly random error
"next" should be relative to the cursor, but I think it's jumping based on some internal ordering or internal caching.

### PR DESCRIPTION
**Pull request by AsaAyers**
Fixes #833

### CODE REVIEW

File: lib/commands.coffee [hunk 1]
```diff
@@ -60,30 +57,49 @@ class Commands
...
-    else
-      @index = 0
-    message = @getMessage(@index)
+    filePath = atom.workspace.getActiveTextEditor().getPath()
+    {row, column} = atom.workspace.getActiveTextEditor().getCursorBufferPosition()
+    thisFile = @linter.views.messages.filter((m) -> filePath is m.filePath )
+
+    message = do ->
+      for m in thisFile
+        {start} = m.range
```
Comment (steelbrain) [inline lib/commands.coffee]: `continue unless m.range`

File: lib/commands.coffee [hunk 2]
```diff
@@ -60,30 +57,49 @@ class Commands
...
+        if start.row > row
+          return m
+        if start.row is row and start.column > column
+          console.log('col', start.column, column)
```
Comment (steelbrain) [inline lib/commands.coffee]: You can just drop this line

File: lib/commands.coffee (hunk 1 above)
Comment (AsaAyers) [inline lib/commands.coffee]: When do messages not have a range? Isn't it required?

Comment (steelbrain) [inline lib/commands.coffee]: Nope, not required.

File: lib/commands.coffee [hunk 3]
```diff
@@ -60,30 +57,48 @@ class Commands
...
     atom.workspace.open(message.filePath).then ->
       atom.workspace.getActiveTextEditor().setCursorBufferPosition(message.range.start)
 
   previousError: ->
-    if @index?
-      @index--
-    else
-      @index = 0
-    message = @getMessage(@index)
+    filePath = atom.workspace.getActiveTextEditor().getPath()
```
Comment (steelbrain) [inline lib/commands.coffee]: `return unless filePath`

File: lib/commands.coffee [hunk 4]
```diff
@@ -60,30 +57,48 @@ class Commands
...
-    # That way the index can be ++ an -- without caring about the array bounds.
-    messages[index %% messages.length]
-
   nextError: ->
-    if @index?
-      @index++
-    else
-      @index = 0
-    message = @getMessage(@index)
+    filePath = atom.workspace.getActiveTextEditor().getPath()
```
Comment (steelbrain) [inline lib/commands.coffee]: `return unless filePath`

**Comment by steelbrain** [comment]
by the way, You don't have to filter messages like that, messages are already classified, You can just do `linter.views.messages.filter((m) => m.currentFile)`

**Comment by AsaAyers** [comment]
ok, that's better than the early return because it maintains the ability to jump to another file that does have an error.

File: lib/commands.coffee [hunk 5]
```diff
@@ -60,30 +57,46 @@ class Commands
...
-    # That way the index can be ++ an -- without caring about the array bounds.
-    messages[index %% messages.length]
-
   nextError: ->
-    if @index?
-      @index++
-    else
-      @index = 0
-    message = @getMessage(@index)
+    {row, column} = atom.workspace.getActiveTextEditor().getCursorBufferPosition()
```
Comment (steelbrain) [inline lib/commands.coffee]: what if `activeTextEditor` is null? It'll throw an `undefined is not a function error`

**Comment by steelbrain** [comment]
Looks good, now all we need is some specs.

**Comment by wmayner** [comment]
+1

**Comment by AsaAyers** [comment]
Specs would be nice, but I'm not really sure how to test this. Can we get this merged without tests?

**Comment by steelbrain** [comment]
I've modified the views and editor-linter classes a bit since this was opened. and You no longer have to filter all of the messages. If you do [`EditorLinter::getMessages()`](https://github.com/atom-community/linter/blob/527d2b3cadf4d9e209cd5d28505560aa3a32d3e7/lib/editor-linter.js#L122-L124) it should be significantly faster than iterating all of the messages. Also the object returned by it is an ES6 Map instead of an array, so you'll have to use the ES6 `for...of` (Coffee's `for...of` would fail) or the `Map.prototype.forEach` function.

**Pull request closed without merging**